tests/bsim/bluetooth/host/att/long_read/main.c

```diff
@@ -0,0 +1,220 @@
+	bt_testlib_bs_sync_all();
+
+	if (get_device_nbr() == 0) {
+		LOG_WRN("Sync point: %s", log_msg);
+	}
+
+	/* Everyone waits for d0 to finish logging. */
+	bt_testlib_bs_sync_all();
+}
+
+static inline void bt_enable_quiet(void)
+{
+	bt_testlib_log_level_set("bt_hci_core", LOG_LEVEL_ERR);
+	bt_testlib_log_level_set("bt_id", LOG_LEVEL_ERR);
+
+	EZ(bt_enable(NULL));
+
+	bt_testlib_log_level_set("bt_hci_core", LOG_LEVEL_INF);
+	bt_testlib_log_level_set("bt_id", LOG_LEVEL_INF);
+}
+
+static void test_long_read(enum bt_att_chan_opt bearer, uint16_t chrc_value_handle,
+			   struct bt_conn *conn)
+{
+	bool central = (get_device_nbr() == 0);
+
+	if (central) {
+		size_t read_count = 0;
+
+		NET_BUF_SIMPLE_DEFINE(attr_value, BT_ATT_MAX_ATTRIBUTE_LEN);
+
+		/* Perform the whole long read operation. */
+		EZ(btt_gatt_long_read(&attr_value, NULL, conn, bearer, chrc_value_handle, 0));
+
+		/* Parse the read attribute value to verify the
+		 * integrity of the transfer.
+		 *
+		 * Each response starts with the length of the whole
+		 * response and the rest is zero-padded.
+		 */
+		for (; attr_value.len; read_count++) {
+			uint16_t encoded_len;
+			uint16_t padding_size;
+
+			LOG_INF("Verifying read %u", read_count);
+
+			__ASSERT(attr_value.len >= sizeof(encoded_len),
+				 "Incomplete encoded length");
+			encoded_len = net_buf_simple_pull_le16(&attr_value);
+
+			padding_size = (encoded_len - 2);
+			LOG_INF("Padding size %u", padding_size);
+
+			/* Check and discard padding. */
+			for (int i = 0; i < padding_size; i++) {
+				__ASSERT(attr_value.len, "Unexpected end of buffer");
+				__ASSERT(net_buf_simple_pull_u8(&attr_value) == 0,
+					 "Expected a padding byte at %u", i);
+			}
+		}
+		LOG_INF("Verified %u reads", read_count);
+		__ASSERT(read_count > 1, "Expected at least two reads");
+	}
+}
+
+void the_test(void)
+{
+	bool central = (get_device_nbr() == 0);
+	bool peripheral = (get_device_nbr() == 1);
+	bt_addr_le_t adva;
+	struct bt_conn *conn = NULL;
+	uint16_t chrc_value_handle = 0;
+
+	if (peripheral) {
+		EZ(bt_gatt_service_register(&svc));
+	}
+
+	bt_enable_quiet();
+
+	if (peripheral) {
+		EZ(bt_set_name("peripheral"));
+		EZ(bt_testlib_adv_conn(&conn, BT_ID_DEFAULT,
+				       (BT_LE_ADV_OPT_USE_NAME | BT_LE_ADV_OPT_FORCE_NAME_IN_AD)));
+	}
+
+	if (central) {
+		EZ(bt_testlib_scan_find_name(&adva, "peripheral"));
+		EZ(bt_testlib_connect(&adva, &conn));
+
+		/* Establish EATT bearers. */
+		EZ(bt_testlib_secure(conn, BT_SECURITY_L2));
+
+		while (bt_eatt_count(conn) == 0) {
+			k_msleep(100);
+		};
+	}
+
+	bs_sync_all_log("Connected");
+
+	/* Perform discovery. */
+	if (central) {
+		find_the_chrc(conn, &chrc_value_handle);
+	}
+
+	bs_sync_all_log("Testing UATT");
+	test_long_read(BT_ATT_CHAN_OPT_UNENHANCED_ONLY, chrc_value_handle, conn);
+
+	bs_sync_all_log("Testing EATT");
+	test_long_read(BT_ATT_CHAN_OPT_ENHANCED_ONLY, chrc_value_handle, conn);
+
+	bs_sync_all_log("Test Complete");
+
+	PASS("Test complete\n");
+}
```

Comment:
I've tried both styles. I ended up writing in this style because it was way easier to program when the interacting code is all in one place. Downside is that the code is limited to be SIMT-like.

Reply:
Alright. I'm in favor of having 2 separate functions, but it is not a blocker.